bitarray/nand.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2014 Workiva, LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bitarray
+
+func nandSparseWithSparseBitArray(sba, other *sparseBitArray) BitArray {
+	// nand is an operation on the incoming array only, so the size will never
+	// be more than the incoming array, regardless of the size of the other
+	max := len(sba.indices)
+	indices := make(uintSlice, 0, max)
+	blocks := make(blocks, 0, max)
+
+	selfIndex := 0
+	otherIndex := 0
+	var resultBlock block
+
+	// move through the array and compare the blocks if they happen to
+	// intersect
+	for {
+		if selfIndex == len(sba.indices) {
+			// The bitarray being operated on is exhausted, so just return
+			break
+		} else if otherIndex == len(other.indices) {
+			// The other array is exhausted. In this case, we assume that we
+			// are calling nand on empty bit arrays, which is the same as just
+			// copying the value in the sba array
+			indices = append(indices, sba.indices[selfIndex])
+			blocks = append(blocks, sba.blocks[selfIndex])
+			selfIndex++
+			continue
+		}
+
+		selfValue := sba.indices[selfIndex]
+		otherValue := other.indices[otherIndex]
+
+		switch {
+		case otherValue < selfValue:
+			// The `sba` bitarray has a block with a index position
+			// greater than us. We want to compare with that block
+			// if possible, so move our `other` index closer to that
+			// block's index.
+			otherIndex++
+
+		case otherValue > selfValue:
+			// Here, the sba array has blocks that the other array doesn't
+			// have. In this case, we just copy exactly the sba array values
+			indices = append(indices, selfValue)
+			blocks = append(blocks, sba.blocks[selfIndex])
+
+			// This is the exact logical inverse of the above case.
+			selfIndex++
+
+		default:
+			// Here, our indices match for both `sba` and `other`.
+			// Time to do the bitwise AND operation and add a block
+			// to our result list if the block has values in it.
+			resultBlock = sba.blocks[selfIndex].nand(other.blocks[otherIndex])
```

Comment:
So I started going down that road at first, but I found that the functions differed beyond just the block comparison. So I tried a way where I passed flags in to do decisioning but then the code got even messier.

As well, I realized that I actually needed to add a new function specifically for nand, to handle the extra case of dense nand sparse. With and, you can handle the dense and sparse case with a single function, since it's commutative, but nand isn't.